CRIU DebugOnRestore mode is disabled when debug events are hooked

JasonFengJ9 opened the pull request:
CRIU `DebugOnRestore` mode is disabled when debug events are hooked

* `isDebugOnRestoreEnabled()` returns `false` if some debug-specific flags/events are enabled/hooked, this matches `J9::Options::isFSDNeeded()` in `compiler/control/J9Options.cpp`;
* `DebugOnRestore` mode is enabled when CRIU is enabled, `-XX:+DebugOnRestore` is specified, JDWP is not enabled, and no debug-specific flags/events are enabled/hooked;
* When `DebugOnRestore` mode is enabled, debug-related events and hooks can't be enabled or hooked pre-checkpoint,
`isDebugAgentDisabled()` returns `TRUE` in this case, otherwise `FALSE`;
* Moved the agent load/start into `criuRestoreInitializeLib()` triggered by `TRIGGER_J9HOOK_VM_PREPARING_FOR_RESTORE`;
* Avoid `AddCapabilities()` for 
`can_generate_single_step_events`/`can_generate_frame_pop_events`/`can_generate_field_modification_events`/`can_generate_field_access_events` such that JIT won't enter `FSD` post-restore if there is no `JDWP` agent specified within the restore option file;
* Renamed `jvmtiHookVMRestoreCRIUInit()` to `jvmtiHookVMPreparingForRestore()`, and `jvmtiHookVMRestoreStartAgent()` to `jvmtiHookVMCRIURestore()` to match the hook trigger events.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

dsouzai commented on runtime/vm/CRIUHelpers.cpp:
```diff
@@ -176,13 +176,37 @@ isJVMInPortableRestoreMode(J9VMThread *currentThread)
+	/* The isDebugEventOrFlagEnabled calculation matches a part of J9::Options::isFSDNeeded()
+	 * in compiler/control/J9Options.cpp.
+	 */
+	bool isDebugEventOrFlagEnabled = J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_FRAME_POPPED)
```
Should this be a `static bool` so that it's only set once? Could the result of this query change over time?

JasonFengJ9 replied:
> Could the result of this query change over time?

```
	bool isDebugEventOrFlagEnabled = J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_FRAME_POPPED)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_FRAME_POP)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_GET_FIELD)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_PUT_FIELD)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_GET_STATIC_FIELD)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_PUT_STATIC_FIELD)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_SINGLE_STEP)
#if defined (J9VM_INTERP_HOT_CODE_REPLACEMENT)
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_POP_FRAMES_INTERRUPT)
#endif
#if defined(J9VM_JIT_FULL_SPEED_DEBUG)
			|| (vm->requiredDebugAttributes & J9VM_DEBUG_ATTRIBUTE_CAN_ACCESS_LOCALS)
#endif
			|| J9_EVENT_IS_HOOKED_OR_RESERVED(vm->hookInterface, J9HOOK_VM_BREAKPOINT);
```
This value could be changed if the event(s) are hooked at runtime by a JVMTI agent or at different stages of VM startup. In addition, the events could be unregistered.
I guess this might affect FSD or debug on restore mode.
Can non-CRIU JVM switch between FSD and default mode, or is it just a one-way direction?

dsouzai replied:
>Can non-CRIU JVM switch between FSD and default mode, or is it just a one-way direction?

No, in a default JVM we're either in FSD mode right from the beginning or we're not.

JasonFengJ9 replied:
> Should this be a static bool so that it's only set once? 

Since this value could change over time, it can't be a static bool.

JasonFengJ9 replied:
Added a field `J9JavaVM->checkpointState.isDebugOnRestoreEnabled` to store the value returned by `isDebugOnRestoreEnabled()`. 
This value is set at the end of VMInitStages `AGENTS_STARTED` and won't be changed afterwards.

This PR is ready for another look.

JasonFengJ9 commented:
Marking this as WIP until a proper way determining the debug on restore mode is found.

tajila commented:
jenkins test sanity alinux jdk21

tajila commented:
jenkins test sanity,extended.functional xlinux jdk17

JasonFengJ9 commented:
> 19:09:28  cmdLineTester_criu_jitPostRestore_3_FAILED

Looking into it.

JasonFengJ9 commented:
Moved the agent load/start into `criuRestoreInitializeLib()` triggered by `TRIGGER_J9HOOK_VM_PREPARING_FOR_RESTORE`;
Avoid `AddCapabilities()` for 
`can_generate_single_step_events`/`can_generate_frame_pop_events`/`can_generate_field_modification_events`/`can_generate_field_access_events` such that JIT won't enter `FSD` post-restore if there is no `JDWP` agent specified within the restore option file;
Renamed `jvmtiHookVMRestoreCRIUInit()` to `jvmtiHookVMPreparingForRestore()`, and `jvmtiHookVMRestoreStartAgent()` to `jvmtiHookVMCRIURestore()` to match the hook trigger events.

This still allows JDWP agents to be enabled post-retore.

@tajila This is ready for another look.

tajila commented:
jenkins test sanity,extended.functional xlinux jdk17
